docs/proposals/PROPOSAL_A.md

```diff
@@ -0,0 +1,68 @@
+# Proposal A - Runtime Annotations
+
+This proposal recommends changes that allow runtimes to quickly select the appropriate image from an index using the platform and annotations on the descriptor.
+
+## Links
+
+| Description | Link |
+|-------------|------|
+| Image Index Spec | <https://github.com/opencontainers/image-spec/blob/main/image-index.md> |
+
+## Modifications
+
+### Image Spec
+
+The following language is already included in the Image Index definition:
+
+> If multiple manifests match a client or runtime's requirements, the first matching entry SHOULD be used.
+
+This would add a section for "Annotations":
+
+> Runtimes MAY be configured to prefer manifests with specific annotations.
+
+An example would also be added to the Image Index:
+
+```jsonc
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "manifests": [
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 7143,
+      "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
+      "platform": {
+        "architecture": "amd64",
+        "os": "linux"
+      }
+    },
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 7682,
+      "digest": "sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270",
+      "platform": {
+        "architecture": "amd64",
+        "os": "linux"
+      },
+      "annotation": {
+        "com.example.gpu": "example-gpu-model"
+      }
+    }
+  ]
+}
+```
+
+In the above scenario, the generic `linux/amd64` entry is listed first and should be used by default.
+Runtimes MAY be configured to prefer the `com.example.gpu: example-gpu-model` entry to run the image customized for their specific environment.
+
+Image authors SHOULD limit annotations to those needed to differentiate descriptors or for runtimes requirements.
+
+Runtimes MAY reject manifests that exceed size limits.
+Runtimes MAY be pre-configured to search for specific annotations matching their feature set.
+Runtimes are not expected to know all the annotations that they prefer for every platform they run on.
+Runtimes MAY have the ability to configure a list of annotations that are preferred by that installation.
```

Comment:
This may add complexity to configuration. For large scale cluster, hosts may be organized into different host aggreggates, each host aggregate may use different generation CPU with different features, then the runtime has to be configured seperately in different host aggregate. If GPU and other factors are included, it become more complex.

Reply:
Requiring runtimes to be aware of every possible 3rd party annotation and update their code to automatically detect it seems unlikely to be approved by the runtime maintainers. The option for administrators to inject some preferred annotations gives flexibility without trying to support every new hardware device and software application in the OCI specs. This does push complexity on to the cluster administrators to inject that configuration data, but since they are already setting up configuration data to provision the nodes, this feels like the right place to add that complexity.

Reply:
>The option for administrators to inject some preferred annotations gives flexibility without trying to support every new hardware device and software application in the OCI specs

if so, administrators can only inject some well known annotations which will be used by image authors. that also means these annotations should be defined in OCI explicitly, otherwise image authors and administrators will use different annotation to represent same meaning.

Reply:
If we only allow OCI defined annotations that puts the burden on us here to manage the space and severely limits different compatibility groups from being empowered to define (and write tools for) their annotation space. I think it’s reasonable for some runtime to select from the space of annotations they care about.

Reply:
@ChaoyiHuang @vsoch I think you both have a very good arguments here. Probably we should figure out some kind of compromise here.

How about trying to define the very familiar and common annotations supported by OCI and leaving room for user definitions?

Reply:
I would say OCI could easily maintain familiar / common annotations (if there is the desire to do that) but we should not place any limitation on what others can do. E.g., remember the table from the early design document that has the [supercontainers annotations](https://docs.google.com/document/d/1nRUuW9i7NRdYXrUr8DQXLMETXwP1J9rw3TbpLTU9yho/edit#bookmark=id.e4ozklypux3s)? It's the runtime (and other tools that use the annotations) that get to decide which to use. Or it could even be passed on by user preference.  That's why nesting might be kind of neat - you could have two equivalent `hardware.gpu.nvidia.driver.version` and instead of having to hard code a specific annotation, tell some tool to use a family instead. E.g., (this is made up):

```
"annotations": {
        "org.network.b": { 
            "mtu": "1350",
            "vxlan": "false",
            ...
    },
        "org.network.b": { 
            "mtu": "3000",
            "vxlan": "true",
             ...
    }
}
```
That's probably a dumb example that means nothing, but it implicitly says "These two orgs are defining attributes for the same thing" and then the user can just say "I want / my container uses org.network.b" (everything under it) instead of a list of the annotations that belong to it. And you can say "I want to compare org.network.a to org.network.b" and (given some common set of annotations for a network provider) you can expect the underlying structure (annotations) to be the same. I'm not sure if it's in scope for our work, but different communities might want to define different sets of common annotation (e.g., for network, things like cpu/gpu, workflow steps) that can be more standardized.

Either way, the rule is the prefix to indicate ownership - org.opencontainers is under our management, and org.supercontainers or `org.<anything else>` is not.

Reply:
> I would say OCI could easily maintain familiar / common annotations (if there is the desire to do that) but we should not place any limitation on what others can do

Yes, I agree with that. My point was only to create common annotations that are not specific to any groups, very generic annotations. At the same time, I don't want to limit users / groups, they could specify their own annotations. The same as you, I want to give them the freedom to express what they want.

> That's why nesting might be kind of neat - you could have two equivalent hardware.gpu.nvidia.driver.version and instead of having to hard code a specific annotation

I am not against the nested structure itself. I am against transforming the `annotations` field (that is a `string-string map`) into a nested structure. That would bring a lot of problems to existing tools.

If we decide to nest, we should most likely create a new compatibility field in the index.
@sudo-bmitch in this proposal suggested to use the `annotations` field. In that case I guess we should create another proposal. 

> Either way, the rule is the prefix to indicate ownership - org.opencontainers is under our management, and org.supercontainers or org.<anything else> is not.

That sounds reasonable to me.